gprs: Resolve MMS proxy/MMSC host name

alfonsosanchezbeato opened the pull request:
Resolve, in case the hostname is provided instead of the IP address, MMS proxy / MMSC name using the DNS servers provided by the context. We need the IP so we can set-up a route for them. Also, use the MMSC for the route in case we do not have a MMS proxy. Finally, export proxy
and proxy port as part of the Settings property array, for both combined and mms context types. Client must use these values to connect to make sure the used IP is the same one as the one for the route, as the MMSC might be a balanced server with multiple IPs.

The PR adds an asynchronous DNS client based on c-ares to avoid blocking ofono while doing the DNS lookup. The client integrates with glib's main loop and has a configurable time-out.

tonyespy commented on configure.ac:
```diff
@@ -188,6 +188,17 @@ AC_SUBST(BLUEZ_CFLAGS)
+				[disable DNS client using c-ares library]),
+						[enable_cares=${enableval}])
+if (test "${enable_cares}" != "no"); then
+	PKG_CHECK_MODULES(CARES, libcares, dummy=yes,
```
The actual library is called libc-ares2.  I actually tried applying as a patch to the ofono bzr tree, and the build failed during override_dh_auto_configure.  I'd added libc-ares-dev to the control file build dependencies, and verified that it was installed in my sbuild chroot, so something's still not right.  Note, I also added --enable_cares to the debian/rules file.  Can you please try a Debian build and report the changes you made?

I would also suggest that you use C_ARES as the Makefile variable, and change the "enable_cares" to "enable-c-ares" as most configure parameters used dashes as opposed to underscores.

alfonsosanchezbeato replied:
As all other settings in the configure file, support for c-ares is on by default, and you can disable it using --disable-cares. So you do not need --enable-cares. enable_cares is a shell name that autoconfigure sets when running "configure" script, and as such needs the underscore (and you can see underscores for these variables in all of configure.ac).

I suggest removing that --enable_cares and trying again, although I will give it a try too.

tonyespy commented on include/dns-client.h:
```diff
@@ -0,0 +1,120 @@
+	OFONO_DNS_CLIENT_ERROR_FORM_ERR,
+	OFONO_DNS_CLIENT_ERROR_SERVER_FAIL,
+	OFONO_DNS_CLIENT_ERROR_NOT_FOUND,
+        OFONO_DNS_CLIENT_ERROR_NOT_IMP,
```
Did you choose to preserve the formatting from the original file?

alfonsosanchezbeato replied:
Yes, but I will keep it only for c-ares-dns-client.c, so I will correct this.

tonyespy commented on include/dns-client.h:
```diff
@@ -0,0 +1,120 @@
+ * defaults
+ * |timeout_ms| specifies a max timeout in milliseconds, or 0 to use the default
+ * timeout.
+ * Returns an opaque handle identifying the request on success and NULL on
```
What's the default timeout?

alfonsosanchezbeato replied:
Seems to be 5 seconds: http://c-ares.haxx.se/ares_init.html

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+ *  DNS Client - provides an asynchronous DNS resolution client.
+ *
+ *  The client is implemented using the c-ares library and integrated with
+ *  flimflam's glib main event loop. See http://c-ares.haxx.se and
```
You might want to update this comment, as the integration is with ofono's glib main event loop, not flimflam's.

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+	if (timeout_provided && timercmp(&elapsed, &request->timeout, >=)) {
+		request->cb(request->data, OFONO_DNS_CLIENT_ERROR_TIMED_OUT,
+			    NULL);
+		stop_ares_request(request);
```
It seems odd that the request is stopped *after* the timeout callback in invoked...

alfonsosanchezbeato replied:
stop_ares_request() adds a callback using g_idle_add(), so it is not really important if you call it after or before request->cb(). It is not clear to me why the deferred destruction is needed though, but seem harmless nonetheless.

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+	if (!read && !write)
+		return TRUE;
+
+	watch = g_malloc0(sizeof(struct ares_watch));
```
Ugh, no NULL check...

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+	ares_destroy(request->channel);
+	g_free(request->hostname);
+	if (request->interface) {
+		rp_filter_set(request->interface, TRUE);
```
It's mentioned above that rp_filter_set doesn't really work...  As it effects the interface as a whole, should we keep this code in place if it's questionable whether or not it works?

alfonsosanchezbeato replied:
I have commented it out, but not removed until I do some further investigation.

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+static gboolean delete_stopped_ares_requests_cb(gpointer data)
+{
+        GList *node, *next;
+        struct ares_request *request;
```
Formatting... ( previous two lines )

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+	}
+
+	if (hostent->h_length > addr_length) {
+		DBG("address too large: %u bytes",
```
Shouldn't some of these errors be error logs vs. DBG?

alfonsosanchezbeato replied:
The CB prints the error too. I have changed those which returned OFONO_DNS_CLIENT_ERROR_NO_DATA as that one can happen for different conditions.

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+}
+
+static void set_request_servers(const char **servers,
+						struct ares_request *request)
```
Formatting?

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+	int rc;
+
+	for (dns = servers; *dns != NULL; ++dns) {
+		current = g_malloc0(sizeof(*current));
```
Not a try...

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+		return NULL;
+	}
+
+	request = g_malloc0(sizeof(struct ares_request));
```
Not a try_...

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+	options.sock_state_cb = ares_socket_state_cb;
+	options.sock_state_cb_data = request;
+	optmask = ARES_OPT_SOCK_STATE_CB;
+	if (timeout_ms > 0) {
```
We should discuss whether or not to keep the original formatting or apply the ofono coding style...

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+		set_request_servers(servers, request);
+		ares_status =
+			ares_set_servers(request->channel, request->servers);
+		/* Log an error, but continue anyway */
```
Why continue?

alfonsosanchezbeato replied:
It is possible that we are able to contact the default DNS servers and we succeed in the end.

tonyespy commented:
I still have more of the review to finish and will plan on doing so first thing tomorrow.  You may want to wait till I've finished before responding to my comments...

alfonsosanchezbeato commented:
A couple of comments before you go on with your review:

1. I have preserved flimflam's formatting. We can adapt it to ofono's standard but I would prefer to keep differences small in case we need to compare.
2. g_malloc0 is what was used originally in flimflam. About its usage, I think we have had this discussion before: I *do prefer* to abort if we are not able to allocate a small number of bytes. If that happens is because something is going very wrong. If it is ofono the one allocating lots of memory, is preferable to abort and let upstart re-start it. If the condition repeats very fast upstart will stop re-starting it too. This is preferable to keep running without the possibility of being able to do any real work because any allocation will not work.

tonyespy commented on src/gprs.c:
```diff
@@ -536,9 +542,228 @@ static void pri_context_signal_settings(struct pri_context *ctx,
+	addr.sin_family = AF_INET;
+	addr.sin_addr.s_addr = INADDR_ANY;
+	memcpy(&rt.rt_genmask, &addr, sizeof(rt.rt_genmask));
+
```
I was going to suggest why you didn't just set addr.sin_family and addr.sin_addr.s_addr once for gateway and netmask, and then I noticed that you did this in delete_route.

I'll point out that most of the code in these two functions is duplicated, so perhaps it'd be better to collapse into a single function that handles adding and removing routes based on a third parameter?

Also, you probably don't need to specify the RTF_UP flag when deleting a route.

tonyespy commented on plugins/c-ares-dns-client.c:
```diff
@@ -0,0 +1,757 @@
+
+/* Intitialize this module. */
+static int c_ares_init(void)
+{
```
I'm assuming the init() and exit() functions are new?  If so, I think it'd be appropriate to follow the ofono ws rules ( blank lines before/after if blocks )... If not, please disregard this comment.

alfonsosanchezbeato replied:
They actually come from chromium (not surprisingly, it seems that connmann/flimflam follows very similar conventions to ofono for plugins).

tonyespy commented on src/dns-client.c:
```diff
@@ -0,0 +1,127 @@
+ *  oFono - Open Source Telephony
+ *
+ *  Copyright (C) 2014  Canonical Ltd.
+ *
```
Does this file require Chromium copyright too?  I ask as there's a comment later in the file that references 'shill', the replacement for 'flimflam'.

alfonsosanchezbeato replied:
Good catch. Right, __ofono_dns_client_strerror comes from chromium.

tonyespy commented:
This looks really good.  I'd added a couple more minor comments...

As for g_malloc0() vs. g_try_malloc0(), it's always hard to look at code that skips a NULL check after allocating memory...  That said, ofono is already inconsistent enough with it's usage of allocation function that it's not worth arguing the point. 

Also two of my original questions were unanswered: (1) what's the default timeout for __ofono_dns_client_submit_request, and (2) why keep the rp_filter code when the associated comments say that it doesn't work?

Finally, my only other comment would be that it would be useful to include a pointer to the repository ( and optionally branch and/or commits ) for the code copied from flimflam/Chromium OS in the respective commit messages.

alfonsosanchezbeato commented:
PR refreshed, addressing comments. See inline answer for the default timeout question. Re: the rp filter I have if-def the code to avoid using it. For the moment I think it's worth keeping the code until we further investigate the issue.

tonyespy commented on src/gprs.c:
```diff
@@ -536,9 +542,181 @@ static void pri_context_signal_settings(struct pri_context *ctx,
+	addr.sin_family = AF_INET;
+	addr.sin_addr.s_addr = inet_addr(ipstr);
+	if (addr.sin_addr.s_addr == INADDR_NONE) {
+		ofono_error("Cannot delete route for invalid IP %s", ipstr);
```
This log message needs an extra string parameter set via: 'create ? "create" : "remove"'.

tonyespy commented on src/gprs.c:
```diff
@@ -536,9 +542,181 @@ static void pri_context_signal_settings(struct pri_context *ctx,
+	memcpy(&rt.rt_genmask, &addr, sizeof(rt.rt_genmask));
+
+	if (ioctl(sk, create ? SIOCADDRT : SIOCDELRT, &rt) < 0)
+		ofono_error("Failed to delete proxy host route: %s (%d)",
```
This log message needs an extra string parameter set via: 'create ? "create" : "remove"'.

tonyespy commented:
Two more minor comments based on the re-factoring of set/delete_route.  Other changes look good.

Also, if dns-client.h and dns-client.c were also taken from flimflam, please add a note in the respective commits as well.

alfonsosanchezbeato commented:
PR refreshed, addressing comments.

tonyespy commented:
Approved.
